packages/discount-code-generator/src/cli.js

```diff
@@ -0,0 +1,227 @@
+    else if (input.path.match(/\.csv$/i))
+      // Convert input stream to CSV file
+      input.pipe(csv({
+        separator: _args.delimiter,
+        strict: true,
+      }))
+        .on('error', (error) => {
+          reject(error)
+        })
+        .on('data', (data) => {
+          const arrayDelim = _args.multivalueDelimiter
+          if (data.cartDiscounts)
+          // eslint-disable-next-line no-param-reassign
+            data.cartDiscounts = data.cartDiscounts.split(arrayDelim)
+          _attributes = data
+        })
+        .on('end', () => {
+          resolve(unflatten(_attributes))
+        })
+    else reject(new Error(
+      'Invalid file format, must be JSON or CSV',
+    ))
+  })
+}
+
+// Resove output to file or stdout
+const resolveOutput = (_args, outputData) => {
+  const outputStream = _args.output
+  const total = outputData.length
+  return new Promise((resolve, reject) => {
+    if (outputStream === process.stdout) {
+      // Write to stdout
+      process.stdout.write(JSON.stringify(outputData, null, 1))
+      resolve(total)
+    } else if (outputStream.path.match(/\.json$/i)) {
+      // Write to json file
+      outputStream.on('error', (error) => {
+        reject(error)
+      })
+      outputStream.end(JSON.stringify(outputData, null, 1))
+      outputStream.on('finish', () => {
+        resolve(total)
+      })
+    } else if (outputStream.path.match(/\.csv$/i)) {
+      // Convert to csv and write to file
+      const arrayDelim = _args.multivalueDelimiter
+      const flatObjects = outputData.map((obj) => {
+        if (obj.cartDiscounts)
+        // eslint-disable-next-line no-param-reassign
+          obj.cartDiscounts = obj.cartDiscounts.join(arrayDelim)
+        return flatten(obj)
+      })
+      const csvOutput = json2csv({
+        data: flatObjects,
+        del: _args.delimiter,
+      })
+      outputStream.on('error', (error) => {
+        reject(error)
+      })
+      outputStream.end(csvOutput)
+      outputStream.on('finish', () => {
+        resolve(total)
+      })
+    } else reject(new Error('Invalid file format. Must be CSV or JSON'))
+  })
+}
+
+// Build discount code options
+const buildOptions = _args => ({
+  quantity: _args['quantity'],
+  length: _args['code-length'],
+  prefix: _args['code-prefix'],
+})
+
+const logError = (error) => {
+  const errorFormatter = new PrettyError()
+
+  if (npmlog.level === 'verbose')
+    process.stderr.write(errorFormatter.render(error))
+  else
+    npmlog.error('', error.message)
+}
+
+const errorHandler = (errors) => {
+  if (Array.isArray(errors))
+    errors.forEach(logError)
+  else
+    logError(errors)
+
+  process.exit(1)
+}
+
+resolveInput(args)
+  .then((attributes) => {
+    const options = buildOptions(args)
+    const codes = discountCodeGenerator(options, attributes)
+    return resolveOutput(args, codes)
+  })
+  .then((total) => {
+    process.stdout.write(`Successfully generated ${total} discount codes\n`)
```

Comment:
This should be done via the logger module, if the user wants the code to be output to _stdout_ inorder to pipe it to another module, this is going to cause an error.

So use the logger to log stuff and the logger should decide where to log either a file or stdout